openlibrary/core/bookshelves.py

```diff
@@ -213,7 +217,34 @@ def get_users_logged_books(
             # XXX Removing limit, offset, etc from data looks like a bug
             # unrelated / not fixing in this PR.
             data = {'username': username}
-        return list(oldb.query(query, vars=data))
+        logged_books = list(oldb.query(query, vars=data))
+
+        # Filtering won't occur unless q >= 3, as limited in mybooks.my_books_view().
+        if q:
+            from openlibrary.plugins.worksearch.code import do_search
+
+            def filter_key(key: str) -> int:
+                """
+                Get the int() representation of an OL work key.
+                filter_key("/works/OL123W")
+                >>> 123
+                """
+                return int("".join(char for char in key if char.isdigit()))
+
+            work_ids = ['/works/OL%sW' % i['work_id'] for i in logged_books]
+            solr_query = {'q': f'{q} key:({" OR ".join(work_ids)})'}
+            solr_resp = do_search(param=solr_query, page=1, rows=100, sort=None)
```

Comment:
Since we're only using this to fetch keys, we can do this to make it speedier:
```suggestion
            solr_resp = run_solr_query(param=solr_query, rows=100, facet=False, fields=['key'])
```